feat(hc): Set up hybrid cloud services for remote calling

RyanSkonnord opened the pull request:
Introduce the RpcModel base class and add a requirement for the Pydantic library. We will use Pydantic to generate OpenAPI schemas for remote procedures and to serialize and deserialize their arguments at runtime.

Introduce the RpcService base class and move metadata from `silo_mode_delegation` calls into the service class declarations. Tag remote procedures with the new `@rpc_method` decorator. Implement local dispatching, with remote dispatching stubbed out.

RyanSkonnord committed:
feat(hc): Set up hybrid cloud services for remote calling

Introduce the RpcModel base class and add a requirement for the Pydantic
library. We will use Pydantic to generate OpenAPI schemas for remote
procedures and to serialize and deserialize their arguments at runtime.

Introduce the RpcService base class and move metadata from
`silo_mode_delegation` calls into the service class declarations. Tag
remote procedures with the new `@rpc_method` decorator. Implement local
dispatching, with remote dispatching stubbed out.

RyanSkonnord commented:
This PR builds on drafts from #44242 and #45200. See also preliminary work in #45281.

If anyone is feeling doubtful about the Pydantic adoption, please note that we are using it as a *parsing tool* for serialization and schema generation. It is not replacing or supplanting MyPy as a *validation tool*.

Regarding @asottile-sentry's [comment](https://github.com/getsentry/sentry/pull/44242#issuecomment-1421282208) about runtime type checking, I think #45281 adequately covers that. This PR would have us begin relying on type annotations for something that we weren't previously, but only in the narrow scope of the declarations of the base services (not even their database-backed implementations), and in a way that will fail conspicuously if it gets messed up.

RyanSkonnord commented on src/sentry/services/hybrid_cloud/app/__init__.py:
```diff
@@ -30,35 +32,32 @@ class RpcSentryAppService:
     sentry_app_id: int = -1
     type: str = ""
-    schema: Mapping[str, Any] = field(default_factory=dict)
+    app_schema: Mapping[str, Any] = Field(default_factory=dict)
```
We need a name change here because `schema` is inherited as a `pydantic.BaseModel` attribute.

RyanSkonnord replied:
Unfortunately no, it's not only a problem with the dynamic model classes. The `RpcSentryAppComponent` class, right here and now, is extending `RpcModel`, which extends `pydantic.BaseModel`. Declaring a `schema` attribute overrides `BaseModel`'s built-in one, and I believe there's a metaclass which detects this and complains.

Personally, I don't think this is worth any special gymnastics. By extending `RpcModel` rather than making simple dataclasses, we inherit methods named
```
"construct", "copy", "dict", "from_orm", "json", "parse_file", "parse_obj",
"parse_raw", "schema", "schema_json", "update_forward_refs", "validate"
```
Several of those, we're already calling elsewhere. We won't have any trouble as long as we treat those as reserved names (and Pydantic will fail early if we forget).

corps replied:
Okie doke

RyanSkonnord committed:
Downgrade pydantic to 1.10.2

For compatibility with
https://github.com/getsentry/pypi/blob/main/packages.ini
(We can change the version there if we need to upgrade.)

getsantry[bot] committed:
:snowflake: re-freeze requirements

getsantry[bot] committed:
:snowflake: re-freeze requirements

RyanSkonnord committed:
Avoid hashing RPC models

RyanSkonnord committed:
Provide non-null default value when serializing team

corps commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -0,0 +1,270 @@
+        if self._base_service_cls.local_mode != SiloMode.REGION:
+            raise RpcServiceSetupException(f"{self.service_name} does not run on the region silo")
+
+        raise NotImplementedError()  # TODO
```
If it would be better for us to actually change signatures and rely on some explicit signal (like always organization_id or region_name), we can accomodate by changing existing services.

RyanSkonnord replied:
I'm exploring this in #45448.

corps replied:
I'll review that next.

corps commented on src/sentry/services/hybrid_cloud/user/impl.py:
```diff
@@ -181,18 +181,16 @@ def serialize_rpc_user(user: User) -> RpcUser:
     args["roles"] = roles
 
-    useremails: FrozenSet[RpcUserEmail] = frozenset({})
+    useremails: List[RpcUserEmail] = []
```
I'm curious why this change was necessary?  Because the actual model does require `frozensets` due to the requirement of `hash` for these models.

RyanSkonnord replied:
Pydantic models, like dataclasses without `(frozen=True, eq=True)`, aren't hashable by default. I get that these `frozenset`s are meant to make `RpcUser` hashable, but they also rely on `RpcUserEmail` being hashable, which isn't either.

They *can* be made hashable (simply by defining a custom `__hash__` method) but I'm experimenting with making it unnecessary. Seems better to go with Pydantic's flow, in that regard. If we actually need logical sets (i.e., discarding duplicate instances of equal objects), it might be simpler and clearer in those cases to explicitly build a dict indexed by ID.

On a philosophical note, I generally like hashable classes better than non-hashable ones, but I also want my hashable classes to be immutable. I vaguely remember a conversation a while back about the advantages of making these models mutable after initialization. Is that still the case?

corps replied:
Absolutely.  IT's kind of a mess.  My ideal world was that the Rpc models had immutable base states and mutable builders, to match the reality of how these classes get used.  That was our older conversation

Part of the trick is that our Rpc classes are often fitting into existing codes and "duck typing" some other type, that may or may not be used in various ways (unfortunately, grabbing a set reference and mutating it alongside doing some work isn't out of the question, or using these classes with hashs as keys or set entries, expecting identifier deduping).  We've often made various decisions to try to imitate the existing behaviors and interfaces of those classes to reduce the impact of our "two code paths" approach (existing SaaS vs incremental region).  So it's been tactical.

I believe that once the migration to 100% regional silo logic is complete, we can untangle this mess by rewriting different parts of the code to remove this.  In the mean time, we have to pick our fights.

That said, I don't have a strong opinion other than we have to be mindful of how these classes are used and find an appropriate solution.  If a custom hash is the best option, I can be convinced of that.

corps replied:
tldr: ok, that's fine by me.  I just want to make sure you're aware and we do something to address the existing usage.  There should, be test coverage.  And you may be seeing those tests fail right now.

corps replied:
Also, as far as "where" models are used as keys...

the answer is:
1.  In notifications code, all over the place and,
2.  In virtually every serializer code path 🙄

RyanSkonnord replied:
Yep, rest assured I'm doing it consciously. I checked `serialize_rpc_user` and there seems to be no risk of a behavioral change there. Not sure about where we care about `RpcUser` being hashed exactly, but I'll follow up if I don't run across it in CI.

corps reviewed (commented):
I definitely like the direction this is going.
My concerns are:
1.  Frozensets are necessary for RpcUser hash-ability, what was the motivating reason for changing the serialization logic to List.
2. Should we escape all parameter field serializations to avoid reserved conflicts?
3. I'd like to pair program with you to start attacking CI failures.
Let me know when and how we can proceed

RyanSkonnord commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -0,0 +1,270 @@
+        name = f"{self.service_name}__{self.method_name}__ParameterModel"
+        parameters = list(inspect.signature(self._base_method).parameters.values())
+        parameters = parameters[1:]  # exclude `self` argument
+        field_definitions = {p.name: create_field(p) for p in parameters}
```
As mentioned above, I think Pydantic would already catch that case and error out. Considering that, I think it's safe enough to just expect developers to avoid those names, and to rely on a fast CI failure if they use one by accident.

corps replied:
My concern is that "schema" is a pretty general, easy to conflict, desirable name.  I do agree fast CI failure is a thing.  I ponder if it might still be worth it to make the interface cleaner by the end.  But it's a nit and not something I would block on as long as we've made an intentional decision for now.

corps replied:
Basically, "ok, sounds good" ;P

markstory commented on src/sentry/services/hybrid_cloud/rpc.py:
```diff
@@ -0,0 +1,270 @@
+
+    @classmethod
+    @abstractmethod
+    def get_local_implementation(cls) -> RpcService:
```
This is a good improvement over how local/remote delegation was wired up before :clap:

markstory reviewed (approved):
Generally looks good to me. We can continue to refine and improve any gaps as we build out the remote dispatch flow.

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Include FilterQueryInterface methods in RPC interface

This adds `serialize_many` and `get_many` as remotely callable methods
for any service that extends FilterQueryInterface.

RyanSkonnord commented on src/sentry/services/hybrid_cloud/filter_query.py:
```diff
@@ -46,6 +49,7 @@ def serialize_many(
     ) -> List[OpaqueSerializedResponse]:
         pass
 
+    @rpc_method
```
@mikejihbe @corps I think this matches the intent of `FilterQueryInterface`, but wanted to be sure.

The effect is that `serialize_many` and `get_many` are automatically included as remotely callable methods on `UserService` and whatever else extends this interface. AFAICT, this is intended and necessary because serializing requires additional database queries from the control silo, so we have no choice but to do it over the wire if it's called on the region silo.

RyanSkonnord committed:
Trying to resolve AttributeError on is_active

(I don't know if this is the correct fix but I want to see what happens
in the tests.)

RyanSkonnord committed:
Replace MayHaveActor with serializable object

RyanSkonnord committed:
Amendment to 37fa929c -- more cautious

RyanSkonnord committed:
Revert changes to src/sentry/auth/superuser.py

That test failure should be fixable by debugging AuthService.

This reverts commit 0a87eebc2b58093c6660bcb26d04805d22f4f9ff.
This reverts commit 37fa929c16b5d0268481b9fc5bc973fdccb713cd.

RyanSkonnord committed:
Have failing remote methods fall back to local impl

RyanSkonnord committed:
Move null check on signature to earlier

So that it's hit properly on region-to-control remote calls.

RyanSkonnord committed:
Change UserIpEvent serialization for logging

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Make RpcOrganizationIntegration.default_auth_id nullable to match ORM model

RyanSkonnord committed:
Simplify recipient issue with IdValueRpcModel

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Fix circular import

RyanSkonnord committed:
Handle nullable webhook_url in serialize_sentry_app

RyanSkonnord commented on src/sentry/services/hybrid_cloud/app/__init__.py:
```diff
@@ -143,9 +154,10 @@ def serialize_sentry_app(self, app: SentryApp) -> RpcSentryApp:
             uuid=app.uuid,
             events=app.events,
-            webhook_url=app.webhook_url,
+            webhook_url=app.webhook_url or "",
```
General question: is this how we want to handle mismatches in nullability between ORM and RPC models?

That is, the ORM `SentryApp` class has

```python
    webhook_url = models.URLField(max_length=512, null=True)
```

but `RpcSentryApp` maps that onto a non-nullable `webhook_url: str = ""`.

Assuming we prefer to represent it that way in the RPC interface, the it makes sense to convert it to the default value during serialization. Unless we want to make the RPC model more faithful to the ORM schema, and change it to `webhook_url: Optional[str] = None`?

corps replied:
That's likely fair.

corps replied:
One partial reason I had originally written this way was, the orm types don't really enforce Nullability in mypy.  When you swap this out, you may?  Have to type fight a little bit.

RyanSkonnord commented:
Extrapolating from https://github.com/getsentry/sentry/pull/45447#discussion_r1139404397, one big (but good!) side effect that maybe we haven't talked about yet, is that a Pydantic model will type-check its attributes at the time the object is initialized, and will raise an error if any of them are wrong.

In other words, assigning `None` to a `webhook_url: str` attribute of a dataclass is no problem, until you unpack it and get an error. After this branch is merged, you'd get an exception on object construction.

RyanSkonnord committed:
Add rpc_method to newly merged method

RyanSkonnord committed:
Add rpc_method dec on update_membership_flags

RyanSkonnord committed:
Terser style for checking if generator is empty

RyanSkonnord committed:
Replace reference to dataclass fields

RyanSkonnord committed:
Add RpcModel.serialize_by_field_name method

Standardize reflective serialization methods.

RyanSkonnord committed:
Simplify RpcModelTest

RyanSkonnord committed:
Require kwargs on ProjectKeyService.get_project_key

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Make RpcActor a hashable RpcModel

RyanSkonnord committed:
Remove IdValueRpcModel

RpcActor replaces the need for RpcUser to be hashed. For other models,
hashability should be handled on a case-by-case basis.

RyanSkonnord committed:
Replace dataclasses.asdict

RyanSkonnord committed:
Experimentally try making RpcUser hashable again

RyanSkonnord committed:
Require kwargs on all service methods; update call sites

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Handle null last_used_at value

RyanSkonnord committed:
Downgrade RpcService error logs

RyanSkonnord committed:
Correct AppService's local mode

Ooooops. :)

RyanSkonnord committed:
Add missing rpc_method on get_custom_alert_rule_actions

RyanSkonnord committed:
Debug _serialize_auth_provider_flags

RyanSkonnord committed:
Fix hybrid_cloud/test_organization.py

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Coerce UUID to string in find_app_components

RyanSkonnord committed:
Tweak _serialize_auth_provider_flags for failing test

getsantry[bot] committed:
:snowflake: re-freeze requirements

RyanSkonnord committed:
Make RpcAuthProvider hashable

RyanSkonnord committed:
Make RpcTeam.org_role nullable for consistency with ORM schema

RyanSkonnord committed:
In TransformTestCase, ensure user has Actor object

Prevents illegal null field when converting to RpcNotificationSetting.

RyanSkonnord committed:
organization_service.get_all_org_roles now requires explicit keyword

RyanSkonnord committed:
Adjust RPC model field nullability to match ORM models

Revert some manual defaults in serializers.

RyanSkonnord committed:
Revert User fields to frozen sets; make sub-objects hashable

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Add test_rpc_model_equals_method

RyanSkonnord committed:
Remove nested frozensets of other RpcModels from RpcUser

As evinced by test_rpc_model_equals_method, Pydantic doesn't seem to
like this.

RyanSkonnord committed:
Minor refactoring in serialize_rpc_user for readability

RyanSkonnord commented:
For those who care, a guided tour of the rabbit hole I recently explored regarding hashability and equality of the RPC model classes...

Believe it or not, [at one point](https://github.com/getsentry/sentry/pull/45447/files#diff-db513c1d5490d3c0bd06abf31ab423fc4d61d8c37936aaee5cc855de86b69aaeR28-R36) I was getting exceptions from using a simple `==` operation on two `RpcUser` objects. It manifested as the following stack trace:

```
pydantic/main.py:909: in pydantic.main.BaseModel.__eq__
pydantic/main.py:450: in pydantic.main.BaseModel.dict
pydantic/main.py:868: in _iter
pydantic/main.py:794: in pydantic.main.BaseModel._get_value
E   TypeError: unhashable type: 'dict'
```

After no small amount of trial, error, and spelunking into the Pydantic code, I determined that it had to do with the `useremails: FrozenSet[RpcUserEmail]` and `authenticators: FrozenSet[RpcAuthenticator]` attributes, and happened whenever one or both was a non-empty set.

We previously ran into some trouble with hashing `RpcModel` objects. So I want to emphasize that the error was **not** caused by anything to do with hashing `RpcUser` itself, nor with hashing `RpcUserEmail` and `RpcAuthenticator`.

Rather, it appears to be a consequence of some of the logic that Pydantic does when comparing two model objects, unpacking the objects' fields and applying some conditional logic based on their types. Something about a frozenset of other Pydantic models causes it to choke.

I never figured out what we might be doing wrong in setting up those fields (other than ensuring that `RpcUserEmail` and `RpcAuthenticator` are properly hashable, which was working correctly at that point) and I'll hesitantly say that this *might* be some kind of bug in Pydantic. I'd be interested to dig into that later, and see if I can isolate the bug under simpler condition.

In the meantime, nested lists of models seem to be happier than frozensets. So [changing those fields to lists](https://github.com/getsentry/sentry/pull/45447/files#diff-5f4678c668f37aa70770743c98d89e4a23c090663e102dcf43004bf80fb48a9fL71-R70) once again. I wanted to explain this in detail, because I was previously exploring the same idea due to difficulties/FUD around hashing the other model classes. This is not that. I actually have some ideas about making our `RpcModel`s more generally hashable. But we can tackle that decision in a later PR, in combination with whether we can make them immutable and require non-default fields at construction time.

For a teaser, here's something fun I learned today about Pydantic. You can drop a nested `Config` class into a Pydantic model like so:

```python
class RpcUserEmail(RpcModel):
    id: int = 0
    email: str = ""
    is_verified: bool = False

    class Config:
        frozen = True
```

And the `frozen` attribute behaves like its `dataclass` equivalent, automatically defining a hash method.

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Add comments and clarify names

RyanSkonnord committed:
Fix typo in docstring

dashed reviewed (approved):
Looks good to me 👍

mikejihbe commented on tests/sentry/hybrid_cloud/test_rpc.py:
```diff
@@ -0,0 +1,48 @@
+
+
+class RpcServiceTest(TestCase):
+    def test_remote_service(self):
```
would also like to see coverage for bad invocations. What happens when you call a non @rpc_method in each region?

RyanSkonnord replied:
Can you elaborate on the kind of bad invocations? Methods not tagged with `@rpc_method` just work normally, and are mostly stuff like serialization helpers.

If you're looking at the `add_organization_member` call below, that was an arbitrary example I picked to exercise the scaffolding. I could add a region-only one to go with it.

If you mean bad invocations like passing serial arguments with bogus parameter names or the wrong types, we won't really be set up to handle those until we've squashed the remaining bugs that trip the error condition [here](https://github.com/getsentry/sentry/blob/62c477285b18b4c59a54ea36488a420e9f7dea87/src/sentry/services/hybrid_cloud/rpc.py#L241-L246). I intend to tackle that when wiring it to an API endpoint, because that's where we'd be receiving the bad invocation and passing back an error response.

RyanSkonnord committed:
Remove TODO comment that's no longer relevant

RyanSkonnord committed:
Merge branch 'master' into rpc-service-wiring

RyanSkonnord committed:
Rename RpcService.name attribute to key
